### ISSUE TITLE
slice & infinite loops

### ISSUE THREAD

**Issue by xgbuils** [benji6/imlazy#18]
Hi!

Implementing lazy `groupBy` method on [iterum](https://github.com/xgbuils/iterum) I noticed that I have some problems with `slice` method with infinite iterables.

I checked that in `imlazy` library there are the same problems. For example:

Given this iterables:
``` javascript
const naturals = range(0)(Infinity)
const naturalsOneTwo = map(e => [...e])([naturals, [1], [2]])
const oneTwoNaturals = map(e => [...e])([[1], [2], naturals])
```

- `slice(Infinity)(Infinity)(naturals)` produces infinite loop but Infinity is not less than Infinity. Then, we can infer that returns empty iterable.

- The same case is `[...slice(5)(2)(naturalsOneTwo)]`. It produces an infinite loop but `5` is greater than `2`. Then, we can infer that returns empty iterable.

- `[...slice(0)(2)(oneTwoNaturals)]` produces an infinite loop but if third element of iterable is not computed the method should return [[1], [2]]. It seems that `slice` computes one more element that requires.

Here my [first attempt](https://github.com/xgbuils/iterum/commit/73cc34bb90bb0150927e81ae0f59cf67b56b4131) to solve these problems. Maybe you can find a implementation less verbose.

**Comment by benji6** [comment]
Hey! I've been quite busy so haven't had time to look at this properly but really appreciate you taking the time to look at it! I tend to favor throwing errors when there is bad user input so might take that approach with the first 2 cases. It might help users realise they are using the function incorrectly more quickly.

The third case is really interesting and I hadn't considered it when writing that code. I see where the error has come in and I'll have a think about a way to fix it soon. I wonder if there are other places in `imlazy` where similar things are happening. I'll have a think about it and hopefully raise a PR in the next few days!

**Comment by xgbuils** [comment]
Hi!

I understand your point about first two cases. However, notice that native javascript method `slice` or `ramda` or `lodash` slice does not have this behaviour and it could be antintuive for the users.

In `haskell` it could even be defined `slice` like:
``` haskell
slice start end = drop start . take end
or
slice start end = take (end - start) . drop start
```

In two definitions, `slice` returns empty list when `start` is greater or equal to `end`.

Cheers!

**Comment by benji6** [comment]
Well maybe you're right about the first 2 cases then. I guess it's easier than adding dynamic type checking to all the functions! I'll probably return an empty iterable then :)

**Comment by benji6** [comment]
Hey I've finally pushed a fix in #19. Found the same issue lurking in my `take` implementation too. Good catch!

**Comment by xgbuils** [comment]
Thanks Ben!

Good job!
